### PR TITLE
doc(monitoring): monitoring guide

### PR DESCRIPTION
**Pull request by bojiang**

### CODE REVIEW

File: docs/source/guides/monitoring.rst [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+Data-Centric Artificial Intelligence (AI) is a popular paradigm that is
+revolutionizing the way we build AI systems. It is a new way of thinking
+about AI that is based on the idea that data is the most important
+component of AI systems.
+
+BentoML embraces this new paradigm by providing APIs that make a data-centric workflow easy to implement.
+
+.. image:: ../_static/img/batching-architecture.png
+
+In this guide, we will focus on the online data collection and model monitoring. BentoML provides a unified interface for that.
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:14]:
```suggestion
This guide will focus on online data collection and model monitoring. We will walk through how one can take full advantage of such features with BentoML. 
```

File: docs/source/guides/monitoring.rst [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+To implement a data collection and model monitoring workflow, you can get these benefits:
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:16]:
```suggestion
The benefits of having a data collection and model monitoring workflow includes:
```

File: docs/source/guides/monitoring.rst [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+- Monitoring key statistical business metrics
+- Identify data drift early to help decide if the model needs to be retrained
+- make QA possible for the previous untracked metrics, like data and model performance
+- use the same data collection and model monitoring workflow for training or model iteration
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:20]:
```suggestion
- Monitor key statistical business metrics.
- Identify early data drift events to determine whether retraining is required.
- Enable QA for the previous untracked metrics, such as model performance, accuracy, degradation, etc.
- Better interoperability for training or model iteration.
```

File: docs/source/guides/monitoring.rst [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+
+Build an ML Application with monitoring API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's take a simple iris classifier service in bentoml/examples as an example below.
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:26]:
```suggestion
The following examples are excerpted from :github:`tree/main/examples/quickstart`.
```

File: docs/source/guides/monitoring.rst [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+Given that we have the definetion of the bento service:
+
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:29]:
```suggestion
```

File: docs/source/guides/monitoring.rst [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+:code:`service.py`
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:30]:
```suggestion
```

File: docs/source/guides/monitoring.rst [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+For some deployment platforms, it's not easy to collect log files. For example, AWS Lambda doesn't support log files. In this case, BentoML can export monitoring data to an OTLP endpoint.
+Some log collectors like fluentbit also supports OTLP input.
+
+
+Plugins and Third-party Monitoring Data Collectors
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+BentoML also supports plugins and third-party monitoring data collectors. User can write a custom monitoring data collector and publish it as a python package.
+Unlike built-ins are more protocol specific for general use cases,
+plugins could be more platform specific.
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:205]:
```suggestion
BentoML supports plugins and third-party monitoring data collectors. Users can implement a custom monitoring data collector.

.. note::

	Currently, in order to use your custom monitoring data collectors, the implementation must be exported as a `Pip package`. See `Python packaging <https://packaging.python.org/en/latest/tutorials/packaging-projects/>`_ for more information on publishing a Python package.
 
Unlike builtins that are protocol-specific for general use cases,
plugins can be implemented to be platform-specific.
```

File: docs/source/guides/monitoring.rst [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+.. note::
+    To use plugins, you need to install the plugin and include it in the dependencies section of the bentofile.
+    For example, it is required to add `bentoml-plugins-arize` to the `python:packages` to use the Arize plugin.
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:209]:
```suggestion
    To use plugins, the plugins need to be enlisted as a dependency under the :ref:`bentofile.yaml <guides/bento:The Build Command>`.

    .. code-block:: yaml
    
    	python:
    	  packages:
    	    - my_monitoring_plugins
```

File: docs/source/guides/monitoring.rst [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+
+1. Ship to Arize AI
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:211]:
```suggestion
1. Arize AI
```

File: docs/source/guides/monitoring.rst [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+For end-to-end solutions for data/model monitoring, BentoML colaborates with Arize AI to provide a plugin for Arize.
+If you don't want to deploy a pipeline by yourself but still need data and model monitoring for the bussiness, Arize AI is a good choice.
+
+Arize AI provides a unified platform for data scientists, data engineers, and ML engineers to monitor, analyze, and debug ML models in production.
+And the `bentoml-plugins-arize` makes it easy to work with BentoML.
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:217]:
```suggestion
Our joint collaboration with `Arize AI <https://arize.com/>`_ offers users an easy-to-use end-to-end deployment (BentoML) to model monitoring (Arize) solution.

Arize AI provides a unified platform for data scientists, data engineers, and ML engineers to monitor, analyze, and debug ML models in production.

BentoML provides a ``bentoml-plugins-arize`` to integrate Arize to BentoML as a plugin.

.. seealso::

	:docs:`integrations/arize` for our integrations with Arize.

.. seealso::

	`Arize Python SDK <https://docs.arize.com/arize/data-ingestion/api-reference/python-sdk/arize.init#keyword-arguments>`_ and `sending real-time logs to Arize <ttps://docs.arize.com/arize/sending-data-to-arize/data-ingestion-methods/sdk-reference/python-sdk/arize.log>`_ for available keyword arguments under ``monitoring.options``.
```

File: docs/source/guides/monitoring.rst [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+
+    monitoring:
+        enable: true
+        type: bentoml_plugins.arize.ArizeMonitor
+        options:
+            space_key: <your_space_key>
+            api_key: <your_api_key>
+            # ... more arize options
+            # see https://docs.arize.com/arize/data-ingestion/api-reference/python-sdk/arize.init#keyword-arguments
+            # and https://docs.arize.com/arize/sending-data-to-arize/data-ingestion-methods/sdk-reference/python-sdk/arize.log
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:231]:
```suggestion
```

File: docs/source/guides/monitoring.rst [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+
+
+
+You could also join the BentoML community and develop your own monitoring data collector by implementing the `bentoml.monitoring.MonitorBase` interface.
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:235]:
```suggestion
.. note::

	To implement your own monitoring data collector, See `bentoml.monitoring.MonitorBase <https://github.com/bentoml/BentoML/blob/d53667da6bedf503e28701d2f0d1c700a5596f64/src/bentoml/_internal/monitoring/api.py>`_ interface.
	
	You can also submit a PR at :github:`bentoml/plugins` to include your monitoring plugin integration to the BentoML ecosystem.
```

File: docs/source/integrations/arize.rst [hunk 13]
```diff
@@ -0,0 +1,41 @@
+========
+Arize AI
+========
+
+Arize AI provides a unified platform for data scientists, data engineers, and ML engineers to monitor, analyze, and debug ML models in production.
+If you don't want to deploy a pipeline by yourself but still need data and model monitoring for the bussiness, Arize AI is a good choice.
```
Comment (aarnphm) [inline docs/source/integrations/arize.rst:6]:
```suggestion
Arize AI provides a unified platform for data scientists, data engineers, and ML engineers to monitor, analyze, and debug ML models in production.
```

File: docs/source/integrations/arize.rst [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+For end-to-end solutions for data/model monitoring, BentoML colaborates with Arize AI to make it easy to work with all BentoML Deployments.
```
Comment (aarnphm) [inline docs/source/integrations/arize.rst:8]:
```suggestion
Our collaboration with Arize AI makes it easy to integrate end-to-end solutions for data/model monitoring with BentoML Deployments.
```

File: docs/source/integrations/arize.rst [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+
+.. seealso::
+   How to use:
+   :ref:`Inference Data Collection & Model Monitoring <guides/monitoring:Plugins and Third-party Monitoring Data Collectors>`.
```
Comment (aarnphm) [inline docs/source/integrations/arize.rst:13]:
```suggestion
.. seealso::
   
   :ref:`Inference Data Collection & Model Monitoring <guides/monitoring:Plugins and Third-party Monitoring Data Collectors>` demonstrates how to use this integration.
```

File: docs/source/integrations/arize.rst [hunk 16]
```diff
@@ -0,0 +1,41 @@
+========
```
Comment (aarnphm) [inline docs/source/integrations/arize.rst:1]: Let's also populate this doc and give a tutorial end-to-end

**Review by aarnphm — COMMENTED** [review]
We definitely need to populate more, but this is a good start.

File: docs/source/guides/monitoring.rst [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+1. Through log files
+
+The most common way to collect monitoring data is to write it to log files. Many utils like fluentbit, filebeat, logstash, etc. can be used to collect log files and ship them to a data warehouse or a monitoring system.
+This is also the default way BentoML exports monitoring data:
+
+.. code:: yaml
+    :caption: `deployment_configuration.yaml`
+
+    monitoring:
+      enable: true
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:171]:
```suggestion
      enabled: true
```

File: docs/source/guides/monitoring.rst (hunk 12 above)
Comment (bojiang) [reply docs/source/guides/monitoring.rst:235]: I'd like to reserve the call of action to join ths community

File: docs/source/guides/monitoring.rst (hunk 1 above)
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:14]:
```suggestion
BentoML embraces this new paradigm by providing a unified API for data-centric workflows.

.. image:: ../_static/img/batching-architecture.png
```

File: docs/source/guides/monitoring.rst [hunk 18]
```diff
@@ -0,0 +1,237 @@
+============================================
+Inference Data Collection & Model Monitoring
+============================================
+
+Data-Centric Artificial Intelligence is a popular paradigm that is
+revolutionizing the way we build AI systems. It is a new way of thinking
+about AI that is based on the idea that data is the most important
+component of AI systems.
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:8]:
```suggestion
Data-centric artificial intelligence is a type of AI that uses data to train and improve its performance. The goal of data-centric AI is to enable machines to process and analyze data in a way that is similar to how humans think and make decisions, allowing them to perform tasks more efficiently and accurately.
```

File: docs/source/guides/monitoring.rst [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+        input=NumpyNdarray.from_sample(np.array([4.9, 3.0, 1.4, 0.2], dtype=np.double)),
+        output=Text(),
+    )
+    async def classify(features: np.ndarray) -> str:
+        results = await iris_clf_runner.predict.async_run([features])
+        result = results[0]
+        category = CLASS_NAMES[result]
+        return category
+
+Before we go to the production, we will only need one more step to add monitoring:
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:58]:
```suggestion
```

File: docs/source/guides/monitoring.rst [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+- make QA possible for the previous untracked metrics, like data and model performance
+- use the same data collection and model monitoring workflow for training or model iteration
+
+
+Build an ML Application with monitoring API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following examples are excerpted from :github:`bentoml/BentoML/tree/main/examples/quickstart`.
+
+Given that we have the definetion of the bento service:
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:30]:
```suggestion
Given the following service definition:
```

File: docs/source/guides/monitoring.rst [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+
+:code:`service.py`
+
+.. code:: python
```
Comment (aarnphm) [inline docs/source/guides/monitoring.rst:34]:
```suggestion
.. code-block:: python
```

File: docs/source/guides/monitoring.rst [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+    async def classify(features: np.ndarray) -> str:
+        results = await iris_clf_runner.predict.async_run([features])
+        result = results[0]
+        category = CLASS_NAMES[result]
+        return category
+
+Before we go to the production, we will only need one more step to add monitoring:
+
+
+.. code:: python
```
Comment (bojiang) [inline docs/source/guides/monitoring.rst:61]:
```suggestion
.. codeblock:: python
```

File: docs/source/guides/monitoring.rst [hunk 23]
```diff
@@ -0,0 +1,238 @@
+============================================
+Inference Data Collection & Model Monitoring
+============================================
+
+Data-Centric Artificial Intelligence is a popular paradigm that is
+revolutionizing the way we build AI systems. It is a new way of thinking
```
Comment (bojiang) [inline docs/source/guides/monitoring.rst:6]:
```suggestion
Data-Centric Artificial Intelligence is an industrial leading paradigm that puts
data at the forefront of AI systems. It is a new way of thinking
```

File: docs/source/guides/monitoring.rst [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+            results = await iris_clf_runner.predict.async_run([features])
+            result = results[0]
+            category = CLASS_NAMES[result]
+
+            mon.log(category, name="pred", role="prediction", data_type="categorical")
+        return category
+
+The Monitor object has a ``log()`` API that allows users to log request features and prediction information. Given data type can be one of the following: ``["numerical", "categorical", "numerical_sequence"]`` with each role to be one of ``["feature", "prediction", "target"]``.
+
+Then we can continue the normal steps, build a bento (BentoML Application) for the above service.
```
Comment (ssheng) [inline docs/source/guides/monitoring.rst:95]: With a complete the service definition, we can proceed to build the bento.

Comment (bojiang) [inline docs/source/guides/monitoring.rst:95]:
```suggestion
With a complete service definition, we can proceed to build the bento.
```

**Review by ssheng — APPROVED** [review]

**Pull request merged**